### PR TITLE
feat: add regression head, loss and transform

### PR DESCRIPTION
**Pull request by lauriebax**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add capability to train regression models.

## Modification

Adds regression heads, losses and single keypoint normalization transform.

## Use cases (Optional)

E.g. train a small resnet-18 regression model:
```yaml
model:
    _delete_: true
    type: 'ImageClassifier'
    backbone:
      type: 'ResNet'
      depth: 18
    neck:
      type: 'GlobalAveragePooling'
    head:
      num_classes: 3 # number of regression values
      type: 'LinearRegressionHead'
      in_channels: 512
      loss:
        type: 'RegressionLoss'
        loss_type: 'L1Loss'
```

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmpretrain/datasets/transforms/singlekeypoint_regression.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+                results['keypoints']
+            ]  # albu expects a list of keypoints, not a single keypoint
+        else:
+            results['keypoints'] = []
+
+        results = self.aug(**results)
+
+        if has_keypoints:
+            if isinstance(results['keypoints'], list):
+                results['keypoints'] = np.array(
```
Comment (Amina-VBTI) [inline mmpretrain/datasets/transforms/singlekeypoint_regression.py:144]: You could put this after line 136, so we don't need to check twice.

File: mmpretrain/datasets/transforms/singlekeypoint_regression.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+                    results['keypoints']).squeeze()  # unpack
```
Comment (Amina-VBTI) [inline mmpretrain/datasets/transforms/singlekeypoint_regression.py:145]: Replace this line with results.pop('keypoints')

File: mmpretrain/models/heads/regression_head.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            feats: The input ``feats`` is a tuple of tensor, and each tensor is
+                the feature of a backbone stage. In ``LinearClsHead``, we just
+                obtain the feature of the last stage.
+        """
+        # The LinearClsHead doesn't have other module, just return after
+        # unpacking.
+        return feats[-1]
+
+    def forward(self, feats: Tuple[torch.Tensor]) -> torch.Tensor:
+        """The forward process."""
```
Comment (Amina-VBTI) [inline mmpretrain/models/heads/regression_head.py:91]: Missing Args

File: mmpretrain/models/heads/regression_head.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                 init_cfg: Optional[dict] = None):
+        if loss is None:
+            loss = dict(type='RegressionLoss',
+                        loss_type='L1Loss',
+                        loss_weight=1.0)
+        super(RegressionHead, self).__init__(init_cfg=init_cfg,
+                                             loss=loss,
+                                             cal_acc=False)
+
+    def _get_predictions(self, cls_score, data_samples):
```
Comment (Amina-VBTI) [inline mmpretrain/models/heads/regression_head.py:30]: Missing Args

File: mmpretrain/datasets/transforms/singlekeypoint_regression.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        self.aug = Compose([self.albu_builder(t) for t in self.transforms],
+                           keypoint_params=keypoint_params)
+
+        if not keymap:
+            self.keymap_to_albu = {'img': 'image'}
+        else:
+            self.keymap_to_albu = keymap
+        self.keymap_back = {v: k for k, v in self.keymap_to_albu.items()}
+
+    def transform(self, results):
```
Comment (Amina-VBTI) [inline mmpretrain/datasets/transforms/singlekeypoint_regression.py:127]: Missing Args

**Review by Jinior — COMMENTED** [review]
Is this new code or did you copy it from one of our existing repositories?
